aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -108,18 +110,217 @@ struct _cpu_scatter_gather_dim_loop {
+  void operator()(
+    scalar_t* self_data, int64_t self_dim_stride,
+    int64_t* index_data, int64_t index_dim_stride,
+    Scalar value,
+    int64_t dim, int64_t index_dim_size,
+    int64_t index_upper_bound,
+    const func_t& f
+  ) {
+
+    for (int64_t i = 0; i < index_dim_size; ++i) {
+      int64_t idx_dim = index_data[i * index_dim_stride];
+      // we are not putting idx_dim in the error message because it disables
+      // loop optimization in clang-7
+      TORCH_CHECK(idx_dim >= 0 && idx_dim < index_upper_bound,
+        "index ", index_data[i * index_dim_stride],
+        " is out of bounds for dimension ", dim,
+        " with size ", index_upper_bound
+      );
+
+      f(
+        self_data + (is_scatter_like ? idx_dim : i) * self_dim_stride,
+        value
+      );
+    }
+  }
 };
 
+// implement reduce_multiply as a class since the multiplication requires a type
+// specialization for the boolean operation (which refuses to proceed the compilation
+// on clang).
+class ReduceMultiply {
+public:
+  ReduceMultiply() {};
+  // don't use auto due to complaints from clang.
+  template <typename scalar_t>
+  void operator()(scalar_t * self_data, scalar_t * src_data) {
+    *self_data *= *src_data;
+  };
+
+  void operator()(bool * self_data, bool * src_data) {
+    *self_data = *self_data && *src_data;
+  };
+};
+ReduceMultiply reduce_multiply;
+
+auto reduce_sum = [](auto * self_data, auto * src_data) {
+                    *self_data += *src_data;
+                  };
+auto reduce_subtract = [](auto * self_data, auto * src_data) {
+                         *self_data -= *src_data;
+                       };
+auto reduce_divide = [](auto * self_data, auto * src_data) {
+                       *self_data /= *src_data;
+                     };
+auto tensor_assign = [](auto * self_data, auto * src_data) {
+                *self_data = *src_data;
+              };
+auto scalar_assign = [](auto * self_data, Scalar src_data) {
+                       using scalar_t = typename std::remove_pointer<decltype(self_data)>::type;
+                       *self_data = src_data.to<scalar_t>();
+                     };
+auto scalar_reduce_add = [](auto * self_data, Scalar src_data) {
+                       using scalar_t = typename std::remove_pointer<decltype(self_data)>::type;
+                       *self_data += src_data.to<scalar_t>();
+                     };
+auto scalar_reduce_subtract = [](auto * self_data, Scalar src_data) {
+                       using scalar_t = typename std::remove_pointer<decltype(self_data)>::type;
+                       *self_data -= src_data.to<scalar_t>();
+                     };
+auto scalar_reduce_multiply = [](auto * self_data, Scalar src_data) {
+                       using scalar_t = typename std::remove_pointer<decltype(self_data)>::type;
+                       *self_data *= src_data.to<scalar_t>();
+                     };
+auto scalar_reduce_divide = [](auto * self_data, Scalar src_data) {
```

Comment:
There is no need for `Scalar` functions if you pass that `Scalar` by its address, then you can reuse your previous callbacks that operate over pointers.